libs/@blockprotocol/graph/src/stdlib/interval.ts

```diff
@@ -0,0 +1,333 @@
+   =========|===========|===========|===========|===========
+   Range A  |    [----] | [--]      | [--]      | [--]
+   Range B  | [-----]   |    [--]   |    (--]   |       (--]
+   ---------|-----------|-----------|-----------|-----------
+   Overlaps |   true    |   true    |   false   |   false
+   */
+  return (
+    (compareBounds(lhs.start, rhs.start, "start", "start") >= 0 &&
+      compareBounds(lhs.start, rhs.end, "start", "end") <= 0) ||
+    (compareBounds(rhs.start, lhs.start, "start", "start") >= 0 &&
+      compareBounds(rhs.start, lhs.end, "start", "end") <= 0)
+  );
+};
+
+/**
+ * Advanced type to provide stronger type information when using {@link intervalIntersectionWithInterval}.
+ *
+ * If *either* of the `start` {@link TemporalBound}s is bounded, then the resultant `start` {@link TemporalBound} will
+ * be bounded, same goes for `end` {@link TemporalBound}s respectively
+ */
+type IntersectionReturn<
+  LhsInterval extends NonNullTimeInterval,
+  RhsInterval extends NonNullTimeInterval,
+> = [LhsInterval, RhsInterval] extends [
+  NonNullTimeInterval<infer LhsStartBound, infer LhsEndBound>,
+  NonNullTimeInterval<infer RhsStartBound, infer RhsEndBound>,
+]
+  ? NonNullTimeInterval<
+      LhsStartBound | RhsStartBound extends TimestampLimitedTemporalBound
+        ? TimestampLimitedTemporalBound
+        : TemporalBound,
+      LhsEndBound | RhsEndBound extends TimestampLimitedTemporalBound
+        ? TimestampLimitedTemporalBound
+        : TemporalBound
+    >
+  : never;
+
+/**
+ * Returns the intersection (overlapping range) of two given {@link NonNullTimeInterval}s, returning `null` if there
+ * isn't any.
+ *
+ * @param {NonNullTimeInterval} lhs
+ * @param {NonNullTimeInterval} rhs
+ */
+export const intervalIntersectionWithInterval = <
+  LhsInterval extends NonNullTimeInterval = NonNullTimeInterval,
+  RhsInterval extends NonNullTimeInterval = NonNullTimeInterval,
+>(
+  lhs: LhsInterval,
+  rhs: RhsInterval,
+): IntersectionReturn<LhsInterval, RhsInterval> | null => {
+  /*
+   Examples     |     1     |     2
+   =============|===========|===========
+   Interval A   |   [-----] | [-----]
+   Interval B   | [-----]   |   [-----]
+   -------------|-----------|-----------
+   Intersection |   [---]   |   [---]
+   */
+  if (!intervalOverlapsInterval(lhs, rhs)) {
+    return null;
+  } else {
+    return {
+      start:
+        compareBounds(lhs.start, rhs.start, "start", "start") <= 0
+          ? rhs.start
+          : lhs.start,
+      end:
+        compareBounds(lhs.end, rhs.end, "end", "end") <= 0 ? lhs.end : rhs.end,
+    } as IntersectionReturn<LhsInterval, RhsInterval>;
+  }
+};
+
+/**
+ * Advanced type to provide stronger type information when using {@link intervalMergeWithInterval}.
+ *
+ * If *both* of the `start` {@link TemporalBound}s are bounded, then the resultant `start` {@link TemporalBound} will be
+ * bounded, same goes for end respectively
+ */
+type MergeReturn<
```

Comment:
@TimDiekmann also starting a comment here so we can thread replies.

> I find the type definitions a bit too complex and don't know if they are needed at all, but I think @CiaranMn is the better person to ask here.

I also had this thought, I admit that there is a risk to maintainability here, but I think they really do help with usability of these functions and this code is probably very unlikely to be touched again much if at all. (And it helps that it's basically all pure TS/JS without any dependencies to introduce exterior reasons for maintenance).

Having these types means that in various places you won't need to check for things like `if (interval.start.kind !== "unbounded")...` and TS doesn't complain if you go directly for `interval.start.limit`.

I did discuss the complexity of these with @thehabbos007 as well, and if I remember correctly, neither of us were _convinced_ that they're worth keeping, but nor was it clear if we should delete them.

They are fairly big types, but I don't think they're that complicated, they're basically a few if statements.

Reply:
> this code is probably very unlikely to be touched again much if at all.

I highly disagree with this reason.

In what situation do you expect to need some kind of `interval.start.kind !== "unbounded"`? The functions are currently only expected to be needed to calculate the intervals, which are used to query data from the graph, which could be any combination of bound types, and I don't see the need for those functions in other places. It feels like an overengineered approach, especially with the statement in mind, that you don't expect to change the code in the future.

Reply:
The _intersection_ of an entity interval with a resolved subgraph interval will necessarily be bounded on both sides for example. This happens throughout subgraph traversal, and subgraph querying.